Plonk: Lagrange basis on demand

vesselinux opened the pull request:
Addresses Issue #55

Commit:
plonk: removed storing the pre-computed Lagrange basis as part of the srs. instead compute it on-the-fly as we need it. addresses issue https://github.com/clearmatics/libsnark/issues/55

Commit:
plonk: added the libfqfft domain parameter as part of the srs. correspondingly function plonk_compute_lagrange_basis was modified to accept domain as an input parameter (from the srs) thus removing the need to compute the domain inside its body.

dtebbs commented:
This PR seems to also contain the commits from #61, so I've made it point to `plonk-transcript-hasher-...` branch so that only the diffs relevant to the Lagrange basis removal show up.  
It would be great if the title also made it clear that this "depends on #61".

Once #61 is merged, this PR will then automatically point to  the `plonk` branch.

dtebbs commented on libsnark/zk_proof_systems/plonk/prover.tcc:
```diff
@@ -713,9 +716,12 @@ round_five_out_t<ppT> plonk_prover<ppT>::round_five(
 
     // --- Computation of r_part[3]
 
+    std::vector<polynomial<Field>> L_basis =
```
Am I right that the Langrage basis is only ever generated in order to use the first element?  If so, why are we generating the entire thing?

If that first polynomial is required, it can  potentially go on the srs.  The idea was to try and not generate them if they are not required, but for things are definitely *are* required, we should mostly avoid generating them multiple times.

vesselinux replied:
Hmm, you make a very good point! 

> Am I right that the Langrage basis is only ever generated in order to use the first element?

Yes, it seems you are right and it is striking that this has escaped me. The only explanation that I have is that the paper is using the full basis e.g. in prover round 2 and I was doing so too in an initial version of the implementation.  Later I replaced the Lagrange basis computation by iFFT in most places and I didn't realize that in the remaining places we are only using the first polynomial from the basis.

Ok, so indeed let's have the first Lagrange polynomial in the srs then! In that case we won't need the domain in the srs anymore. Would you recommend to still keep it in case we may need it in the future (although premature optimization is generally not a good thing I guess)?

vesselinux replied:
Added the 0-th Lagrange basis polynomial in the SRS. Not computing the full basis on demand any more. The `domain` is still part of the SRS for now.

vesselinux commented:
> This PR seems to also contain the commits from #61, so I've made it point to `plonk-transcript-hasher-...` branch so that only the diffs relevant to the Lagrange basis removal show up. It would be great if the title also made it clear that this "depends on #61".
> 
> Once #61 is merged, this PR will then automatically point to the `plonk` branch.

Rebased on `plonk` now and dropped commits relevant to #61 so that only the relevant files appear in the diffs.

Commit:
plonk: added the 0-th Lagrange basis polynomial in the SRS instead of computing the full basis on demand (we only use the 0-th polynomial from the basis); addresses PR comment https://github.com/clearmatics/libsnark/pull/62#discussion_r933200581 .

dtebbs commented on libsnark/zk_proof_systems/plonk/utils.tcc:
```diff
@@ -42,22 +44,20 @@ template<typename FieldT> void print_vector(const std::vector<FieldT> &v)
 template<typename FieldT>
-void plonk_compute_lagrange_basis(
-    const size_t npoints, std::vector<polynomial<FieldT>> &L)
+std::vector<polynomial<FieldT>> plonk_compute_lagrange_basis(
```
Is this still required?
```console
$ git grep plonk_compute_lagrange_basis
libsnark/zk_proof_systems/plonk/srs.tcc:    // see \ref plonk_compute_lagrange_basis()
libsnark/zk_proof_systems/plonk/utils.hpp:std::vector<polynomial<FieldT>> plonk_compute_lagrange_basis(
libsnark/zk_proof_systems/plonk/utils.tcc:std::vector<polynomial<FieldT>> plonk_compute_lagrange_basis(
```

vesselinux replied:
No, it is not used indeed. Removed.

Commit:
plonk: removed unused function plonk_compute_lagrange_basis()

dtebbs reviewed (approved):
LGTM

dtebbs commented:
@vesselinux seems to need rebasing onto origin/plonk

vesselinux commented:
> @vesselinux seems to need rebasing onto origin/plonk

Done

dtebbs commented:
Do we really not need the domain anymore? I would expect we'd need it for
the FFT and iFFT.

I guess you can try to remove it but that would be surprising if it were
unused.

On Fri, Jul 29, 2022, 23:46 Vesselin Velichkov ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In libsnark/zk_proof_systems/plonk/prover.tcc
> <https://github.com/clearmatics/libsnark/pull/62#discussion_r933347992>:
>
> > @@ -713,9 +716,12 @@ round_five_out_t<ppT> plonk_prover<ppT>::round_five(
>
>      // --- Computation of r_part[3]
>
> +    std::vector<polynomial<Field>> L_basis =
>
> Hmm, you make a very good point!
>
> Am I right that the Langrage basis is only ever generated in order to use
> the first element?
>
> Yes, it seems you are right and it is striking that this has escaped me.
> The only explanation that I have is that the paper is using the full basis
> e.g. in prover round 2 and I was doing so too in an initial version of the
> implementation. Later I replaced the Lagrange basis computation by iFFT in
> most places and I didn't realize that in the remaining places we are only
> using the first polynomial from the basis.
>
> Ok, so indeed let's have the first Lagrange polynomial in the srs then! In
> that case we won't need the domain in the srs anymore. Would you recommend
> to still keep it in case we may need it in the future (although premature
> optimization is generally not a good thing I guess)?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/clearmatics/libsnark/pull/62#discussion_r933347992>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADIKWIKHWFTSGF4UUNJMXTVWPVEDANCNFSM54WA44CA>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>
